### ISSUE TITLE
My tinyproxy as reverse proxy is working very slow 

### ISSUE THREAD

**Issue by pavelsr** [tinyproxy/tinyproxy#198]
Hello,

I use tinyproxy as reverse proxy for making web interface of 3G modem connected to raspberry pi available on LAN address of RPi. 

Ip address of RPi in LAN is `192.168.0.106` and IP address of RPi that was assigned by DHCP server of modem is `192.168.8.100`. I want that page available at `192.168.8.100` become available via `192.168.0.106` 

Reverse proxying is working, but after proxying web interface of modem through tinyproxy html is loading very slow and I can't access other webpages except main.

Here is my configuration ( `/etc/tinyproxy/tinyproxy.conf` ):

```
User tinyproxy
Group tinyproxy
Port 80
Timeout 60
DefaultErrorFile "/usr/share/tinyproxy/default.html"
StatFile "/usr/share/tinyproxy/stats.html"
Logfile "/var/log/tinyproxy/tinyproxy.log"
LogLevel Info
PidFile "/run/tinyproxy/tinyproxy.pid"
MaxClients 50
MinSpareServers 30
MaxSpareServers 50
StartServers 50
MaxRequestsPerChild 0
ViaProxyName "tinyproxy"
ConnectPort 80
ConnectPort 443
ConnectPort 563
# upstream 192.168.8.1:80
ReversePath "/" "http://192.168.8.1/"
ReverseOnly Yes
ReverseMagic Yes
ReverseBaseURL "http://192.168.0.106/"
````
[Here](https://paste.ubuntu.com/p/cRTsdRbXV2/) is log of openning web page.

I noticed that message 

```
INFO      Sep 15 09:09:58 [14791]: No upstream proxy for 192.168.8.1
```

is constantly repeating. 

But connection is closing immediately if I uncomment `# upstream 192.168.8.1:80` in config:

```
CONNECT   Sep 15 09:07:01 [14677]: Connect (file descriptor 7): 192.168.0.110 [192.168.0.110]
CONNECT   Sep 15 09:07:01 [14677]: Request (file descriptor 7): GET /html/home.html HTTP/1.1
CONNECT   Sep 15 09:07:01 [14677]: Rewriting URL: /html/home.html -> http://192.168.8.1/html/home.html
INFO      Sep 15 09:07:01 [14677]: Found upstream proxy 192.168.8.1:80 for 192.168.8.1
CONNECT   Sep 15 09:07:01 [14677]: Established connection to upstream proxy "192.168.8.1" using file descriptor 8.
INFO      Sep 15 09:07:01 [14677]: Closed connection between local client (fd:7) and remote client (fd:8)
```

What could be wrong in my configuratuion?

Version: `tinyproxy 1.8.4`

```
Features compiled in:
    XTinyproxy header
    Filtering
    Transparent proxy support
    Reverse proxy support
    Upstream proxy support
```

**Comment by rofl0r** [comment]
maybe try portredir instead ( https://github.com/rofl0r/portredir )

**Comment by pavelsr** [comment]
It wasn't helpful

When I run `sudo redir --laddr=192.168.0.106 --lport=80 --caddr=192.168.8.1 --cport=80`user agent just go to `192.168.8.1`.

```
curl -vL http://192.168.0.106/
*   Trying 192.168.0.106...
* Connected to 192.168.0.106 (192.168.0.106) port 80 (#0)
> GET / HTTP/1.1
> Host: 192.168.0.106
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 307 Temporary Redirect
< Date: Thu, 01 Jan 1970 00:00:00 GMT
< Server: WebServer
< Connection: close
< X-Download-Options: noopen
< X-Frame-Options: deny
< X-XSS-Protection: 1; mode=block
< Strict-Transport-Security: max-age=31536000; includeSubdomains
< Location: http://192.168.8.1/html/index.html?url=192.168.0.106
< Content-Length:13
< Cache-Control: no-cache
< Content-Type: text/html
< 
* Closing connection 0
* Issue another request to this URL: 'http://192.168.8.1/html/index.html?url=192.168.0.106'
*   Trying 192.168.8.1...
```

But this address is on another interface and network, of course it will not be accessible! 
That's why I need to setup reverse proxy.

Interesting thing that nginx also does not open page, though http statuses are all 200. [Here](https://paste.ubuntu.com/p/KfNzn8tYcd/) is nginx log FYI.

**Comment by rofl0r** [comment]
looks as if that site insists on being addressed with the raw ip address it knows about itself...

**Comment by anphsw** [comment]
Hello!
I noticed same problem in my uclibc/mips build.
Each request just delayed for 15 seconds in function getpeer_information.
Maybe this patch 8a52f6776b2b6f73f3e91ccd8372224038f6c952 will broke something in another code, but with it reverse proxy works without delays.

**Comment by rofl0r** [comment]
@anphsw so is your patch there fixing build issues ? if so, what is the exact error you're getting ? and how is it related to slowness issues reported here ?

**Comment by anphsw** [comment]
@rofl0r not build issues, but issues when building against mips/uclibc.

Pages opened very slowly.
Exact error:
1. Browser requests page from tinyproxy
2. Process has accepted connection (same is in log: "Process has accepted connection")
3. exactly 15 seconds of nothing (getpeer_information is somehow broken)
4. "Connect (file descriptor..." in logs
5. Request is processed normally.
This error exist only in embedded build, x86 works fine.
Toolchain is "crosstools-mips-gcc-4.6-linux-3.4-uclibc-0.9.32-binutils-2.21" and target is Sercomm RV6699 router.

Maybe it's not exact same error, but it's may be releated to this topic, better to try build and test it.

**Comment by rofl0r** [comment]
> better to try build and test it

ok, please send me your mips device and toolchain! (j/k)

> exactly 15 seconds of nothing (getpeer_information is somehow broken)

could you do me a favor and run tinyproxy under strace `strace -f -s 256 -o logfile.txt tinyproxy --foo --bar` and after like 5 seconds of that timeout happening, hit CTRL-C and supply me the log ? you can find a precompiled strace binary for mips [here](http://landley.net/aboriginal/downloads/old/binaries/1.2.4/extras/)

**Comment by anphsw** [comment]
Done.
Archive contents:
1. console debug log
2. strace log
3. config file
4. build flags

[tinyproxy-bug-log.tar.gz](https://github.com/tinyproxy/tinyproxy/files/2717709/tinyproxy-bug-log.tar.gz)

**Comment by rofl0r** [comment]
that's with getpeer_information re-enabled, right ?
i can't really figure out what's going on, maybe it could help if you add a debug line before and after the getpeer_information call

**Comment by anphsw** [comment]
Yes, it's with enabled getpeer_information.
I added debug prints and what is happening:

ERROR     Jan 01 00:59:09 [14586]: get_ip_string finished
NOTICE    Jan 01 00:59:10 [14584]: Waiting servers (1) is less than MinSpareServers (2). Creating new child.
DEBUG     Jan 01 00:59:10 [14584]: [child.c:498] INC: servers_waiting: 2
{malloc: 0x446bd8:128} child.c:195
ERROR     Jan 01 00:59:24 [14586]: getnameinfo finished (0)

So it seems like getnameinfo returns success, but waits for a long time.
I tried other utils using getnameinfo, such as nslookup and ping, but they work fine and quick.

**Comment by rofl0r** [comment]
ok i get it. basically tinyproxy tries to do a DNS resolve for the ip connecting to it to display a "nice" line in the log. that seems utterly stupid. working on a fix.

**Comment by rofl0r** [comment]
i added a fix for the issue in the `threads` branch ^
i would appreciate if you could test it.

**Comment by anphsw** [comment]
Your fix works. Thank you!

**Comment by rofl0r** [comment]
thanks for testing! does the tinyproxy from the `threads` branch generally work well for you, @anphsw ? IIRC uclibc's threading code is a bit weak

**Comment by anphsw** [comment]
I using only reverse proxy function, a it works well on threads branch. I dont noticed any differences during testing.

**Comment by michael-doubez** [comment]
Is @rofl0r branch merged yet ? The performance hit (due to getnameinfo timeout) was a deal breaker for me.

**Comment by rofl0r** [comment]
no, the `threads` branch is still in the "please test" phase.

**Comment by mshgh** [comment]
just my 2c
I run into the same problem on `Alpine 3.9.3 x86_64` with `tinyproxy-1.10.0-r1`
Adding IP address and hostname of the calling PC into `/etc/hosts` on the tinyproxy box made it running as fast as expected.
Sure not a fix, but I think in some cases this can be acceptable workaround. Was for me ;)

**Comment by rofl0r** [comment]
fix was merged to master as f6d4da5d81694721bf50b2275621e7ce84e6da30
